### ISSUE TITLE
Misc. bug: performance regression in `llama-server` (`ggml-vulkan`)

### ISSUE THREAD

**Issue by kpouget** [ggml-org/llama.cpp#17033]
### Name and Version

master branch

### Operating systems

Mac

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
llama-server --model /path/to/llama3.2 --host 0.0.0.0 --port 11434 --n-gpu-layers 99
```

### Problem description & steps to reproduce

 I think https://github.com/ggml-org/llama.cpp/pull/16736 introduced a regression in the performance of `llama-server`, when backed by `ggml-vulkan` (but no regression with `ggml-metal`, which seems strange):

* on [2025-11-02](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-psap-topsail-main-mac_ai-jump-ci-cpt/1984953945801363456/artifacts/jump-ci-cpt/005-test/artifacts/test-artifacts/004__plots/regression_report.html) `b6923`
    * TTFT `3777ms`
    * ITL `41ms` 

* on [2025-11-03](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-psap-topsail-main-mac_ai-jump-ci-cpt/1985316382480273408/artifacts/jump-ci-cpt/005-test/artifacts/test-artifacts/004__plots/regression_report.html) `b6933`
    * TTFT `7459ms`
    * ITL `62ms`

Note that `llama-bench` isn't affected, `pp512=223t/s` and `tg128=28t/s` didn't degrade.

I looked further at the performance of the commits between these two days (tests ran on another system, so the absolute values are different)

`b6924..b6927`

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/ff7b2fdb-4e3a-4d48-a82f-5ab29c1c1868" />

```
    - b6924
    - cd5e3b575 # b6925
    - 2f966b8ed # b6926
    - b6927
 ```

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/d55ecfac-92cb-4e46-b671-1d3842409654" />

and they highlight that the performance dropped occurred with cd5e3b575, when https://github.com/ggml-org/llama.cpp/pull/16736 was merged.

### First Bad Commit

cd5e3b575

### Relevant log output

```shell

```

**Comment by ggerganov** [comment]
How do you determine TTFT in your tests?

**Comment by kpouget** [comment]
```
# First non empty token is the first token
result.first_token_time = tokens[0]['time']
```
from what I see [there](https://github.com/openshift-psap/llm-load-test/blob/main/src/llm_load_test/plugins/openai_plugin.py#L356-L357), it's the time between the request (`start_time`) and the first non empty token.

The load tester extracts the stats about all the TTFT, and my test uses the median value for the regression testing
(I can lookup the rest of the TTFT / ITL stats if that can be useful)

**Comment by ggerganov** [comment]
If you add `-kvu` to your command does it restore the TTFT?

**Comment by kpouget** [comment]
yes indeed, with `-kvu` the TTFT and ITL performance are back to normal

* `b6924` is before the regression
* `b6927` is after the regression
* both tests ran with the `-kvu` flag

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/c6b5141e-52b6-412a-80d9-ff4ec476d23a" />

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/ff765856-74ff-4465-a5b9-673cfe2c79f3" />

**Comment by ggerganov** [comment]
Since this happens with the Vulkan backend and does not happen with the Metal backend, I think it is because the Vulkan FA implementation does not have the optimization for skipping -INF blocks (https://github.com/ggml-org/llama.cpp/pull/16372). @jeffbolznv can you confirm this is the case?

Since by default now the server uses 4 slots with unified cache (#16736) this is likely resulting in the regression. You can use `-kvu -np 1` which would result in the old single-slot behavior where this optimization is not necessary.

**Comment by jeffbolznv** [comment]
Right, ggml-vulkan does not yet skip -INF blocks. Can you clarify what has changed about the shape of the tensors we'll see and what shape of -INF blocks we should be looking for?

Is it possible to reproduce this change in behavior in llama-bench?

**Comment by kpouget** [comment]
> Is it possible to reproduce this change in behavior in llama-bench?

if you're asking if the performance degradation is visible in `llama-bench`, according to my tests it's not, the perf didn't change

> Note that llama-bench isn't affected, `pp512=223t/s` and `tg128=28t/s` didn't degrade.

**Comment by ggerganov** [comment]
> Can you clarify what has changed about the shape of the tensors we'll see and what shape of -INF blocks we should be looking for?

When using the unified KV cache, the different sequences are end up in the same buffer. For example, with 4 slots, when we don't use the unified KV cache, the KV cache buffers can look like this:

```bash
# . - empty cell
# N - sequence id

00000000...................
11111111111111111111.......
2222.......................
333333333333...............
```

With unified KV cache, the buffer is single-stream and can look like this:

```bash
00000000111111111111111111112222333333333333..............................
```

So now, if you are processing a new token for seq `2` for example, the mask would look like this:

```bash
# . - masked value (i.e. -INF)

............................0000............
```

I.e. it attends only to the tokens of sequence `2`. But in contrast to the non-unified KV cache, they are not necessarily at the beginning of the buffer.

In the Metal backend, I've made an optimization where I run a quick kernel to detect blocks in the KQ mask full of -INF and then during the FA kernel, simply skip them.

Note that while in this case this is close to optimal, it still has some drawbacks. For example, the unified KV cache could end up looking like this in some cases:

```bash
0123013012301230123012301230123012301230123012301230...........................
```

So now, the logic for searching -INF blocks would not help because every block would likely end up having at least one non masked token. Noting this since there might be better approaches than https://github.com/ggml-org/llama.cpp/pull/16372 (e.g. creating a list of indices).

> Is it possible to reproduce this change in behavior in llama-bench?

Currently no. It seems we don't have a good way to bench this case. We should figure out something. It might be easier to add an option in `llama-batched-bench` for that.

**Comment by jeffbolznv** [comment]
Is it possible to reproduce this with llama-cli? Or only with llama-server?

**Comment by ggerganov** [comment]
No, llama-cli does not support multiple sequences.

**Comment by jeffbolznv** [comment]
@kpouget can you share detailed instructions for how to reproduce this? I haven't really used llama-server.

**Comment by kpouget** [comment]
```
$ llama-server --model /path/to/llama3.2 --host 0.0.0.0 --port 11434 --n-gpu-layers 99
```
```
$ git clone https://github.com/openshift-psap/llm-load-test
$ cd llm-load-test
$ pip install -r ./requirements.txt
```
<details>
<summary>config.yaml</summary>

```
output:
  format: "json"
  dir: "./output"
  file: "output.json"
storage:
  type: local
dataset:
  file: "datasets/openorca_large_subset_011.jsonl"
  max_queries: 3000
  min_input_tokens: 0
  min_output_tokens: 0
  max_input_tokens: 2047
  max_output_tokens: 1024
  max_sequence_tokens: 2048
load_options:
  type: constant #Future options: loadgen, stair-step
  concurrency: 1
  duration: 30
plugin: "openai_plugin"
plugin_options:
  streaming: True
  interface: "http"
  model_name: "/path/to/llama3.2" # change this
  host: "http://localhost:11434"
  endpoint: "/v1/chat/completions"
  use_tls: False
extra_metadata:
  replicas: 1
```
</details>

```
# update plugin_options.model_name to match the llama-server model name:
$ curl -Ssf localhost:11434/models | jq -r .models[0].name
```
```
$ python3 load_test.py --config  ./config.yaml
```
```
$ cat output/output.json | jq .summary.ttft.median
90.32535552978516
```
let me know if this works for you

**Comment by jeffbolznv** [comment]
Thanks, I was able to get this running. I see it report about 74 vs 58 with -kvu, on RTX 5090. These seem orders of magnitude different from what's in your original chart, but similar enough to your last comment, so maybe this is expected.

**Comment by kpouget** [comment]
the plots in my top post are from a Macbook M4 Pro. Is the RTX 5090 so much faster than that? (I don't know)
ggml-vulkan is only 75%-80% of ggml-metal, but your figures are on a different level

**Comment by jeffbolznv** [comment]
I also measured on RTX 4070 with coopmat1/2 disabled and I get 232 vs 335. I logged the FA sizes and see that before the change, the KV cache is smaller and its size resets periodically. After the change, it grows larger and ends up being around KV==3500 at the end. I guess this is in line with what Georgi described. I'll look into doing the mask optimization when I get a chance.

**Comment by jeffbolznv** [comment]
@kpouget could you try this commit: https://github.com/jeffbolznv/llama.cpp/commit/e25bf37a7617741f475c9e380b344b492d068aef. I just did the check in the shader, I got 241 on RTX 4070.

**Comment by ggerganov** [comment]
> It might be easier to add an option in llama-batched-bench for that.

@jeffbolznv I've extended `llama-batched-bench` to support this use case: https://github.com/ggml-org/llama.cpp/pull/17103. LMK if it is not clear how to use it.

**Comment by jeffbolznv** [comment]
Thanks. I ran the new test and it does show an improvement with the commit I shared. I still see slowdown as the number of sequences increases, and I actually still see that slowdown even if I completely nop out the FA work in ggml-vulkan. Not sure what's going on there.

I'd like to hear back from @kpouget before I finish up this change.

**Comment by kpouget** [comment]
> @kpouget could you try this commit: https://github.com/jeffbolznv/llama.cpp/commit/e25bf37a7617741f475c9e380b344b492d068aef.

sure, I'll test that first thing Monday morning!

**Comment by ggerganov** [comment]
> I still see slowdown as the number of sequences increases, and I actually still see that slowdown even if I completely nop out the FA work in ggml-vulkan. Not sure what's going on there.

At large context, the creation of the KQ mask in host memory can make a difference. This is best mitigated by making the backend asynchronous so that `llama_kv_cache::set_input_kq_mask()` would get interleaved with the GPU work.

**Comment by jeffbolznv** [comment]
Thanks for the tip. I've prototyped this and I get a speedup on the first iteration, but after the first iteration I see extra calls to backend_synchronize coming from ggml_backend_sched_alloc_splits. Not sure what's going on, I'll debug it.

**Comment by jeffbolznv** [comment]
The strange behavior is caused by graph_optimize. I think what's happening is: there's a "warmup" that sizes the KV cache to the full size, the first round of prompt processing (graph_optimize does nothing) uses this full size cache and needs no reallocation, then some token gen happens and graph_optimize changes the node order and causes a reallocation, then when we get to the second set of prompt processing the cache grows on each iteration and needs reallocation for every ubatch of prompt processing (the cache grows 1MB per iteration).

Not sure what the right fix is. I guess real-world usage could have a similar issue when alternating between pp and tg?

**Comment by ggerganov** [comment]
And with graph_optimize disabled you don't see this behavior?

**Comment by jeffbolznv** [comment]
Right.

**Comment by jeffbolznv** [comment]
I tried moving node_alloc into hash_node, rather than having the array galloc->node_allocs. That seems like it would be resilient to the graph order changing. I get a different behavior - the first test iteration reallocates over and over, the later iterations don't (and are fast).

It looks like the graph is being rebuilt for each ubatch during the first iteration. I guess the existing logic could reuse allocations across different graphs (was this intended)? Should this test be reusing the graph in the first iteration?

**Comment by kpouget** [comment]
So, comparing apples to apples, here is what I had before 

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/2c477eba-af69-4223-b741-36f11db5d640" />

and now: `TTFT=834.88ms`

<img width="1200" height="650" alt="Image" src="https://github.com/user-attachments/assets/fc2c96d4-ce0c-4acb-a47c-c2a2d574a396" />

and now: `ITL=17.64ms`

so compared to the before the regression, 
* the TTFT improved 
* the ITL degraded of 1-2ms

**Comment by jeffbolznv** [comment]
Thanks for testing. If the KV cache is bigger now then I think a small decrease in token generation perf is probably expected given the current implementation.

**Comment by kpouget** [comment]
the overall throughput slightly decreased, [from](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift-psap_topsail/862/pull-ci-openshift-psap-topsail-main-mac_ai-jump-ci/1986350663914754048/artifacts/jump-ci/005-test/artifacts/test-artifacts/000__matbenchmarking/mac_ai/000__test/004__llm_load_test__run/output/output.json)
>  38.71041090839531

to 
> 36.842331020929144

however I get some crashes from time to time:
```
srv  get_availabl: prompt cache update took 12.69 ms
.../llama.cpp/tools/server/server.cpp:3843: pos_min == -1, but n_past > 0 - should not happen: https://github.com/ggml-org/llama.cpp/pull/13833#discussion_r2116181237
slot launch_slot_: id  2 | task -1 | sampler chain: logits -> logit-bias -> penalties -> dry -> top-n-sigma -> top-k -> typical -> top-p -> min-p -> xtc -> temp-ext -> dist
slot launch_slot_: id  2 | task 3196 | processing task
slot update_slots: id  2 | task 3196 | new prompt, n_ctx_slot = 4096, n_keep = 0, task.n_tokens = 808
slot update_slots: id  2 | task 3196 | n_past = 5, slot.prompt.tokens.size() = 151, seq_id = 2, pos_min = -1
```
I don't think it was there before.
Let me know if you think it's related and need further info about the crash, I'll recompile with debug symbols
```
    frame #1: 0x00000001901c7848 libsystem_pthread.dylib`pthread_kill + 296
    frame #2: 0x00000001900d09e4 libsystem_c.dylib`abort + 124
    frame #3: 0x000000010063d470 libggml-base.dylib`ggml_abort + 160
    frame #4: 0x000000010009a65c llama-server`server_context::update_slots() + 15664
    frame #5: 0x000000010000aa0c llama-server`server_queue::start_loop() + 760
    frame #6: 0x0000000100007d54 llama-server`main + 20240
```

**Comment by ggerganov** [comment]
These crashes should be fixed with https://github.com/ggml-org/llama.cpp/commit/cb1adf885105da7ce23db746b4202f4e987aa3e8
